File: VB/WebApiLogClient.vb
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.IO
Imports System.Net
Imports DevExpress.Utils
Imports DevExpress.Xpo.Logger
Imports DevExpress.Xpo.Logger.Transport
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq

Namespace XpoProfilerConsole
	Friend Class WebApiLogClient
		Implements ILogClient

		Private host As Host
		Private Shared ReadOnly timeSpanConverter As New TimeSpanJsonConverter()
		Public Property Login() As String
		Public Property Password() As String
		Public Property AuthToken() As String
		Public Sub New(ByVal host As Host)
			Me.host = host
		End Sub
		Public Sub New(ByVal host As Host, ByVal login As String, ByVal password As String, ByVal token As String)
			Me.host = host
			Me.Login = login
			Me.Password = password
			Me.AuthToken = token
		End Sub
		Private Function GetServiceUrl() As String
			Dim protocol As String = ""
			If Not host.HostAddress.StartsWithInvariantCultureIgnoreCase("http://") AndAlso Not host.HostAddress.StartsWithInvariantCultureIgnoreCase("https://") Then
				protocol = "http://"
			End If
			Return String.Format(CultureInfo.InvariantCulture, "{0}{1}:{2}/{3}/", protocol, host.HostAddress, host.Port, host.ServiceName)
		End Function
		Private Function GetHttpResponse(ByVal url As String) As String
			Dim request As HttpWebRequest = CType(HttpWebRequest.Create(url), HttpWebRequest)
			If Not String.IsNullOrEmpty(Login) Then
				request.Credentials = New NetworkCredential(Login, Password)
			End If
			If Not String.IsNullOrEmpty(AuthToken) Then
				request.Headers.Add("Authorization", String.Concat("XpoProfilerToken ", AuthToken))
			End If
			Dim response As HttpWebResponse = CType(request.GetResponse(), HttpWebResponse)
			Using str As New StreamReader(response.GetResponseStream())
				Return str.ReadToEnd()
			End Using
		End Function
		Public Function GetCompleteLog() As LogMessage() Implements DevExpress.Xpo.Logger.Transport.ILogSource.GetCompleteLog
			Dim url As String = String.Concat(GetServiceUrl(), "GetCompleteLog")
			Dim json As String = GetHttpResponse(url)
			Return JsonConvert.DeserializeObject(Of LogMessage())(json, timeSpanConverter)
		End Function
		Public Function GetMessage() As LogMessage Implements DevExpress.Xpo.Logger.Transport.ILogSource.GetMessage
			Dim url As String = String.Concat(GetServiceUrl(), "GetMessage")
			Dim json As String = GetHttpResponse(url)
			Return JsonConvert.DeserializeObject(Of LogMessage)(json, timeSpanConverter)
		End Function
		Public Function GetMessages(ByVal messagesAmount As Integer) As LogMessage() Implements DevExpress.Xpo.Logger.Transport.ILogSource.GetMessages
			Dim url As String = String.Concat(GetServiceUrl(), "GetCompleteLog?messagesAmount=", messagesAmount.ToString())
			Dim json As String = GetHttpResponse(url)
			Return JsonConvert.DeserializeObject(Of LogMessage())(json, timeSpanConverter)
		End Function
		Public Function Start() As Boolean Implements ILogClient.Start
			Return True
		End Function
		Public Function [Stop]() As Boolean Implements ILogClient.Stop
			Return True
		End Function
		Public Sub Dispose() Implements System.IDisposable.Dispose
		End Sub
	End Class

	Friend Class TimeSpanJsonConverter
		Inherits JsonConverter

		Public Overrides Function CanConvert(ByVal objectType As Type) As Boolean
			Return objectType Is GetType(TimeSpan)
		End Function
		Public Overrides Function ReadJson(ByVal reader As JsonReader, ByVal objectType As Type, ByVal existingValue As Object, ByVal serializer As JsonSerializer) As Object
			Select Case reader.TokenType
				Case JsonToken.String
					Dim val As String = reader.Value.ToString()
					Dim ts As TimeSpan = Nothing
					If TimeSpan.TryParse(val, CultureInfo.InvariantCulture, ts) Then
						Return ts
					End If
					Throw New InvalidOperationException(String.Format("Cannot deserialize TimeSpan from string '{0}'.", val))
				Case JsonToken.StartObject
					Dim jObject As JObject = JObject.Load(reader)
					Dim ticks As JToken = jObject("ticks")
					If ticks IsNot Nothing Then
						Return New TimeSpan(ticks.ToObject(Of Integer)())
					End If
					Throw New InvalidOperationException(String.Format("Cannot deserialize TimeSpan from JSON '{0}'.", jObject))
			End Select
			Throw New InvalidOperationException(String.Format("Cannot deserialize TimeSpan from token '{0}'", reader.TokenType))
		End Function
		Public Overrides Sub WriteJson(ByVal writer As JsonWriter, ByVal value As Object, ByVal serializer As JsonSerializer)
			Throw New NotImplementedException()
		End Sub
	End Class
End Namespace

File: CS/WebApiLogClient.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Net;
using DevExpress.Utils;
using DevExpress.Xpo.Logger;
using DevExpress.Xpo.Logger.Transport;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace XpoProfilerConsole {
    internal class WebApiLogClient : ILogClient {
        Host host;
        static readonly TimeSpanJsonConverter timeSpanConverter = new TimeSpanJsonConverter();
        public string Login { get; set; }
        public string Password { get; set; }
        public string AuthToken { get; set; }
        public WebApiLogClient(Host host) {
            this.host = host;
        }
        public WebApiLogClient(Host host, string login, string password, string token) {
            this.host = host;
            this.Login = login;
            this.Password = password;
            this.AuthToken = token;
        }
        string GetServiceUrl() {
            string protocol = "";
            if(!host.HostAddress.StartsWithInvariantCultureIgnoreCase("http://")
                && !host.HostAddress.StartsWithInvariantCultureIgnoreCase("https://")) {
                protocol = "http://";
            };
            return string.Format(CultureInfo.InvariantCulture, "{0}{1}:{2}/{3}/", protocol, host.HostAddress, host.Port, host.ServiceName);
        }
        string GetHttpResponse(string url) {
            HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(url);
            if(!string.IsNullOrEmpty(Login)) {
                request.Credentials = new NetworkCredential(Login, Password);
            }
            if(!string.IsNullOrEmpty(AuthToken)) {
                request.Headers.Add("Authorization", string.Concat("XpoProfilerToken ", AuthToken));
            }
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            using(StreamReader str = new StreamReader(response.GetResponseStream())) {
                return str.ReadToEnd();
            }
        }
        public LogMessage[] GetCompleteLog() {
            string url = string.Concat(GetServiceUrl(), "GetCompleteLog");
            string json = GetHttpResponse(url);
            return JsonConvert.DeserializeObject<LogMessage[]>(json, timeSpanConverter);
        }
        public LogMessage GetMessage() {
            string url = string.Concat(GetServiceUrl(), "GetMessage");
            string json = GetHttpResponse(url);
            return JsonConvert.DeserializeObject<LogMessage>(json, timeSpanConverter);
        }
        public LogMessage[] GetMessages(int messagesAmount) {
            string url = string.Concat(GetServiceUrl(), "GetCompleteLog?messagesAmount=", messagesAmount.ToString());
            string json = GetHttpResponse(url);
            return JsonConvert.DeserializeObject<LogMessage[]>(json, timeSpanConverter);
        }
        public bool Start() {
            return true;
        }
        public bool Stop() {
            return true;
        }
        public void Dispose() {
        }
    }

    class TimeSpanJsonConverter : JsonConverter {
        public override bool CanConvert(Type objectType) {
            return objectType == typeof(TimeSpan);
        }
        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer) {
            switch(reader.TokenType) {
                case JsonToken.String:
                    string val = reader.Value.ToString();
                    TimeSpan ts;
                    if(TimeSpan.TryParse(val, CultureInfo.InvariantCulture, out ts)) {
                        return ts;
                    }
                    throw new InvalidOperationException(string.Format("Cannot deserialize TimeSpan from string '{0}'.", val));
                case JsonToken.StartObject:
                    JObject jObject = JObject.Load(reader);
                    JToken ticks = jObject["ticks"];
                    if(ticks != null) {
                        return new TimeSpan(ticks.ToObject<int>());
                    };
                    throw new InvalidOperationException(string.Format("Cannot deserialize TimeSpan from JSON '{0}'.", jObject));
            }
            throw new InvalidOperationException(string.Format("Cannot deserialize TimeSpan from token '{0}'", reader.TokenType));
        }
        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer) {
            throw new NotImplementedException();
        }
    }
}
